java-checks/src/main/java/org/sonar/java/checks/maven/PomElementOrderCheck.java

```diff
@@ -0,0 +1,123 @@
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.LinkedList;
+import java.util.List;
+
+@Rule(
+  key = "S3423",
+  name = "pom elements should be in the recommended order",
+  priority = Priority.MINOR,
+  tags = {Tag.CONVENTION, Tag.MAVEN})
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.READABILITY)
+@SqaleConstantRemediation("10min")
+public class PomElementOrderCheck implements MavenFileScanner {
+
+  @Override
+  public void scanFile(MavenFileScannerContext context) {
+    MavenProject project = context.getMavenProject();
+    List<Location> issues = checkPositions(
+      project.getModelVersion(),
+      project.getParent(),
+      project.getGroupId(),
+      project.getArtifactId(),
+      project.getVersion(),
+      project.getPackaging(),
+      project.getName(),
+      project.getUrl(),
+      project.getInceptionYear(),
+      project.getOrganization(),
+      project.getLicenses(),
+      project.getDevelopers(),
+      project.getContributors(),
+      project.getMailingLists(),
+      project.getPrerequisites(),
+      project.getModules(),
+      project.getScm(),
+      project.getIssueManagement(),
+      project.getCiManagement(),
+      project.getDistributionManagement(),
+      project.getProperties(),
+      project.getDependencyManagement(),
+      project.getDependencies(),
+      project.getRepositories(),
+      project.getPluginRepositories(),
+      project.getBuild(),
+      project.getReporting(),
+      project.getProfiles()
+      );
+
+    if (!issues.isEmpty()) {
+      context.reportIssue(this, project.startLocation().line(), "Reorder the elements of this pom to match the recommended order.", issues);
+    }
+  }
+
+  private static List<Location> checkPositions(LocatedTree... trees) {
+    List<Location> issues = new LinkedList<>();
+    List<LocatedTree> expectedOrder = getNonNullTrees(trees);
+    List<LocatedTree> observedOrder = sortByLine(expectedOrder);
+
+    for (int index = 0; index < expectedOrder.size(); index++) {
+      LocatedTree expected = expectedOrder.get(index);
+      int indexObserved = observedOrder.indexOf(expected);
+      if (!issues.isEmpty() || index != indexObserved) {
+        issues.add(new Location("Expected position: " + (index + 1), expected));
+      }
+    }
+
+    return issues;
+  }
+
+  private static List<LocatedTree> getNonNullTrees(LocatedTree... trees) {
+    List<LocatedTree> result = new LinkedList<>();
+    for (LocatedTree locatedTree : trees) {
+      if (locatedTree != null) {
+        result.add(locatedTree);
+      }
+    }
+    return result;
+  }
+
+  private static List<LocatedTree> sortByLine(List<LocatedTree> expectedOrder) {
+    List<LocatedTree> result = new LinkedList<>(expectedOrder);
+    Collections.sort(result, new Comparator<LocatedTree>() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this a named "static" inner class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2694)